xcodeproj/repositories.bzl

```diff
@@ -0,0 +1,66 @@
+"""Definitions for handling Bazel repositories used by rules_xcodeproj."""
+
+load(
+    "@com_github_buildbuddy_io_rules_xcodeproj//xcodeproj/internal:logging.bzl",
+    "green",
+    "yellow",
+    "warn",
+)
+
+def _maybe(repo_rule, name, ignore_version_differences, **kwargs):
+    """Executes the given repository rule if it hasn't been executed already.
+
+    Args:
+        repo_rule: The repository rule to be executed (e.g., `http_archive`.)
+        name: The name of the repository to be defined by the rule.
+        ignore_version_differences: If `True`, warnings about potentially
+            incompatible versions of depended-upon repositories will be
+            silenced.
+        **kwargs: Additional arguments passed directly to the repository rule.
+    """
+    if native.existing_rule(name):
+        if not ignore_version_differences:
+            # Verify that the repository is being loaded from the same URL and
+            # tag that we asked for, and warn if they differ.
+            # This isn't perfect, because the user could load from the same
+            # commit SHA as the tag, or load from an HTTP archive instead of a
+            # Git repository, but this is a good first step toward validating.
+            # Bzlmod will remove the need for all of this in the longer term.
+            existing_repo = native.existing_rule(name)
+            if (existing_repo.get("remote") != kwargs.get("remote") or
+                existing_repo.get("tag") != kwargs.get("tag")):
+                expected = "{url} (tag {tag})".format(
+                    tag = kwargs.get("tag"),
+                    url = kwargs.get("remote"),
+                )
+                existing = "{url} (tag {tag})".format(
+                    tag = existing_repo.get("tag"),
+                    url = existing_repo.get("remote"),
+                )
+
+                warn("""\
+`com_github_buildbuddy_io_rules_xcodeproj` depends on `{repo}` loaded from \
```

Comment:
nit: do you want to define a common "name" var somewhere and reference everywhere to ease the move of this into another repo down the line?

Reply:
I don't think we could use that in `load` or `workspace` statements, and besides those this is the only place we spell it out. Should be easy enough to do the replacement when we do a move.